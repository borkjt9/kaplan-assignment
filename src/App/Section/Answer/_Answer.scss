$option-height: 60px;
$option-border-radius: 10px;
$option-background: #F9F9F9;
$placeholder-primary-color: lightgray;
$border-color: lightgray;
$shadow-color: #595959;

.answer {
  position: relative;
  background-color: inherit;
  width: 100%;
  &__answer-btn {
    width: inherit;
    min-height: $option-height;
    background: $option-background;
    border: 1px solid $border-color;
    border-radius: $option-border-radius;
    p {
      font-size: 1rem;
    }
    &.--active {
      width: 104%;
      margin-left: -2%;
      background: #007bff;
      box-shadow: 1px 1px 3px $border-color;
      color: white;
    }
     &:focus {
       outline: none;

     }
  }

  &__arrow-btn {
    position: absolute;
    padding: 0px;
    background: inherit;
    border: none;

    top: 0;
    &.--hor{
      height: $option-height;
      img {
        height: calc(#{$option-height}  - 20px);
      }
    }

    &.--left {
      right: calc(-34%);
    }
    &.--right {
      left: calc(-34%);

    }
    &:focus {
      outline: none;
    }

    img {
      width: 30px;
    }

  }

  &__arrow-btns {
    position: absolute;
    right: calc(-34% + 30px);
    height: $option-height;
    display: block;
    top: 0px;
    &__arrow-btn {
      height: calc(#{$option-height} / 2 - 5px);
      padding: 0px;
      background: inherit;
      border: none;
      display: block;
      &.--vert {

      }

      &:focus {
        outline: none;
      }

      &.--up {
        margin-bottom: 10px;
        img {
          height: inherit;

        }
      }

      &.--is-disabled {
        img {
          -webkit-filter: grayscale(100%) brightness(150%); /* Safari 6.0 - 9.0 */
          filter: grayscale(100%) brightness(150%);
        }
        opacity: 75%;
      }

      &.--down {
        margin-top: 10px;
        img {
          height: inherit;
        }
      }
    }
  }
}
