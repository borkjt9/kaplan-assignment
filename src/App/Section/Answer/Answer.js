import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { DragSource } from 'react-dnd';
import { dragPreviewSrc, horArrowSrc, upArrowSrc, downArrowSrc } from '../../../shared/assets/assetURLs';
import { DRAGGABLE_TYPE_ANSWER } from '../../../redux/constants';
import './_Answer.scss';

/**
 * Used as a parameter to DragSource. Which is a higher order component from react-dnd library.
 * http://react-dnd.github.io/react-dnd/docs-overview.html
 */
const answerSource = {
  /**
   * Tells DragSource to begin drag.
   * @param {Object} props - The components props, plus added props from react-dnd.
   */
  beginDrag(props) {
    const {
      id,
      left,
      top,
      side,
      index,
    } = props;

    props.setActiveAnswer({ side, index });
    return { id, left, top };
  },
};

/**
 * From react-dnd documentation:
 * For each component that needs to track the drag and drop state,
 * You can define a collecting function that retrieves the relevant bits of it from the monitors.
 * React DnD then takes care of timely calling your collecting function
 * and merging its return value into your components' props.
 */
function collect(connect, monitor) {
  return {
    connectDragSource: connect.dragSource(),
    connectDragPreview: connect.dragPreview(),
    isDragging: monitor.isDragging(),
  };
}

class Answer extends Component {
  /**
   * Answers are rendered in locations whenever an answer is not.
   * Connected to react-dnd higher order component 'DropTarget'.
   */

  componentDidMount() {
    // Supplies img to be used in drag action.
    const img = new Image();
    img.src = dragPreviewSrc;
    img.style.opacity = 0.5;
    this.props.connectDragPreview(img);
  }

  /**
   * Called when either of the arrow buttons are clicked.
   *
   * @param {Number} toIndex - New location of answer.
   * @param {Number} fromIndex - Old location of answer.
   * @param {String} dir - Either 'hor' or 'vert'. used to set toSide and FromSide.
   */
  callPositionAnswer(toIndex, fromIndex, dir) {
    const { positionAnswer, activeQuestion, side } = this.props;
    const data = {
      toIndex,
      fromIndex,
      activeQuestion,
    };
    if (side === 'left' && dir === 'hor') {
      data.toSide = 'right';
      data.fromSide = 'left';
    } else if (side === 'left' && dir === 'ver') {
      data.toSide = 'left';
      data.fromSide = 'left';
    } else if (side === 'right' && dir === 'hor') {
      data.toSide = 'left';
      data.fromSide = 'right';
    } else {
      data.toSide = 'right';
      data.fromSide = 'right';
    }
    positionAnswer(data);
  }

  /**
   * Called when either of the arrow buttons are clicked.
   *
   * @param {Number} index - Used to disable up and down buttons if at bottom or top of answers.
   */
  renderArrowButtons(index) {
    const { side, totalAnswers } = this.props;
    const btnHorizClassName = `answer__arrow-btn --hor --${side}`;
    const horizBtn = () => (
      <button onClick={() => this.callPositionAnswer(index, index, 'hor')} className={btnHorizClassName}>
        <img
          className="answer__arrow-btn__img"
          alt="move answers to left side or right side"
          src={horArrowSrc}
        />
      </button>
    );
    if (side === 'left') {
      return (
        horizBtn()
      );
    }
    let btnUpClassName = `answer__arrow-btns__arrow-btn --vert --up --${side}`;
    if (index === 0) {
      btnUpClassName += ' --is-disabled';
    }
    const upBtn = () => (
      <button
        onClick={() => this.callPositionAnswer(index - 1, index, 'vert')}
        className={btnUpClassName}
      >
        <img
          className="answer__arrow-btns__arrow-btn__up__img"
          alt="move answers to left side or right side"
          src={upArrowSrc}
        />
      </button>
    );

    let btnDownClassName = `answer__arrow-btns__arrow-btn --vert --down --${side}`;
    if (index === totalAnswers - 1) {
      btnDownClassName += ' --is-disabled';
    }
    const downBtn = () => (
      <button
        onClick={() => this.callPositionAnswer(index + 1, index, 'vert')}
        className={btnDownClassName}
      >
        <img
          className="answer__arrow-btns__arrow-btn__down__img"
          alt="move answers to left side or right side"
          src={downArrowSrc}
        />
      </button>
    );
    return (
      <div>
        {horizBtn()}
        <div className="answer__arrow-btns --vert">
          {upBtn()}
          {downBtn()}
        </div>
      </div>
    );
  }

  render() {
    const {
      connectDragSource,
      isDragging,
      toggleArrowButtons,
      activeAnswer,
      side,
      index,
      answersDict,
    } = this.props;
    let answerBtnClassName = 'answer__answer-btn';
    if (activeAnswer.index === index && activeAnswer.side === side) {
      answerBtnClassName += ' --active';
    }

    //connectDragSource is react-dnd function.
    return connectDragSource(
      <div
        action={this.action}
        className="answer my-4"
      >
        <button onClick={() => toggleArrowButtons(index)} className={isDragging ? `${answerBtnClassName}  --active` : answerBtnClassName}>
          <p className="m-2">{answersDict[index]}</p>
        </button>
        { (activeAnswer.index === index && activeAnswer.side === side) ? this.renderArrowButtons(index) : ''}
      </div>
    );
  }
}

Answer.propTypes = {
  /** Supplied by react-dnd. */
  connectDragSource: PropTypes.func.isRequired,
  /** Supplied by react-dnd. Sets position of hover item. */
  connectDragPreview: PropTypes.func.isRequired,
  /** Supplied by react-dnd. True if drag item is being dragged. */
  isDragging: PropTypes.bool.isRequired,
  /** activates / deactives answer. All makes arrow buttons visible. */
  toggleArrowButtons: PropTypes.func.isRequired,
  /**
   * Current active answer. Supplied to toggleArrowButtons.
   * EXAMPLE: {side: 'left', index: 2}
   */
  activeAnswer: PropTypes.objectOf(PropTypes.string).isRequired,
  /** Side answer is found on. */
  side: PropTypes.string.isRequired,
  /** current index of answer. */
  index: PropTypes.number.isRequired,
  /**
   * Total answers for the question.
   * Used to render total number of answers and placeholders.
   */
  answersDict: PropTypes.objectOf(PropTypes.any).isRequired,
  /**
   * Total answers for the question.
   */
  totalAnswers: PropTypes.number.isRequired,
  /** Changes location of answer in section. */
  positionAnswer: PropTypes.func.isRequired,
  /** Active question number. */
  activeQuestion: PropTypes.number.isRequired,
};

export default DragSource(DRAGGABLE_TYPE_ANSWER, answerSource, collect)(Answer);
