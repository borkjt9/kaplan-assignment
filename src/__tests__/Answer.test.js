import React from 'react';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import renderer from 'react-test-renderer';
import { Answer } from '../App/Section/Answer/Answer';

Enzyme.configure({ adapter: new Adapter() });
const identity = el => el;


const answersDictDummy = {
  0: 'Test',
  1: 'Hi',
  3: 'Hello',
};

describe('Answer', () => {
  let wrapper;

  it('renders snapshot correctly', () => {
    const renderedValue =  renderer.create(
      <Answer
        toggleArrowButtons={identity}
        activeAnswer={{ side: 'left', index: 0 }}
        side="left"
        index={0}
        answersDict={answersDictDummy}
        totalAnswers={5}
        positionAnswer={identity}
        activeQuestion={0}
        setActiveAnswer={identity}
        connectDragSource={identity}
        connectDragPreview={identity}
        isDragging
      />).toJSON();
    expect(renderedValue).toMatchSnapshot();
  });

  it('renders correct text', () => {
      wrapper = shallow(
        <Answer
          toggleArrowButtons={identity}
          activeAnswer={{ side: 'left', index: 0 }}
          side="left"
          index={0}
          answersDict={answersDictDummy}
          totalAnswers={5}
          positionAnswer={identity}
          activeQuestion={0}
          setActiveAnswer={identity}
          connectDragSource={identity}
          connectDragPreview={identity}
          isDragging

        />);
    const renderedText = wrapper.text();
    const expectedText = answersDictDummy[0];
    expect(renderedText).toEqual(expectedText);
  });

  it('arrows-hor-button renders on left side active answer', () => {
    wrapper = shallow(
      <Answer
        toggleArrowButtons={identity}
        activeAnswer={{ side: 'left', index: 0 }}
        side="left"
        index={0}
        answersDict={answersDictDummy}
        totalAnswers={5}
        positionAnswer={identity}
        activeQuestion={0}
        setActiveAnswer={identity}
        connectDragSource={identity}
        connectDragPreview={identity}
        isDragging

      />);
    const buttons = wrapper.find('button');
    expect(buttons.length).toEqual(2);
  });

  it('arrows-hor-button and vert-buttons render on right side active answer', () => {
    wrapper = shallow(
      <Answer
        toggleArrowButtons={identity}
        activeAnswer={{ side: 'right', index: 0 }}
        side="right"
        index={0}
        answersDict={answersDictDummy}
        totalAnswers={5}
        positionAnswer={identity}
        activeQuestion={0}
        setActiveAnswer={identity}
        connectDragSource={identity}
        connectDragPreview={identity}
        isDragging

      />);
    const buttons = wrapper.find('button');
    expect(buttons.length).toEqual(4);
  });

  it('no arrow buttons render on right side not active answer', () => {
    wrapper = shallow(
      <Answer
        toggleArrowButtons={identity}
        activeAnswer={{ side: 'right', index: 0 }}
        side="right"
        index={3}
        answersDict={answersDictDummy}
        totalAnswers={5}
        positionAnswer={identity}
        activeQuestion={0}
        setActiveAnswer={identity}
        connectDragSource={identity}
        connectDragPreview={identity}
        isDragging

      />);
    const buttons = wrapper.find('button');
    expect(buttons.length).toEqual(1);
  });
});
