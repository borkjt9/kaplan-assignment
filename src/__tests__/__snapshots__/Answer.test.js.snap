// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Answer renders snapshot correctly 1`] = `
<div
  action={undefined}
  className="answer my-4"
>
  <button
    className="answer__answer-btn --active  --active"
    onClick={[Function]}
  >
    <p
      className="m-2"
    >
      Test
    </p>
  </button>
  <button
    className="answer__arrow-btn --hor --left"
    onClick={[Function]}
  >
    <img
      alt="move answers to left side or right side"
      className="answer__arrow-btn__img"
      src="https://firebasestorage.googleapis.com/v0/b/kaplan-assignment.appspot.com/o/horizontal-arrows.svg?alt=media&token=c4ea3d7b-5327-429b-9909-3b26f946c24e"
    />
  </button>
</div>
`;
